### PR TITLE
refactor: migrate to TypeScript

### PR DESCRIPTION
**Pull request by LordFoogThe4rd**
## Summary

Migrate the entire codebase (frontend and server) from JavaScript to TypeScript. Enables strict mode and adds comprehensive type definitions across all modules, eliminating `as any` casts and improving type safety.

### TypeScript Migration
- **Frontend**: Convert all `.js`/`.jsx` source files to `.ts`/`.tsx` — including `App`, `AppLayout`, all components (Sidebar, Modals, Widgets, controls), all API modules (openai, koboldcpp, llamacpp, aihorde, deepseek, common), all hooks (`useGenerationLogic`, `useKeyboardShortcuts`, `usePersistentState`, `usePromptBuilder`, `useScreenshotCapture`, `useSessionState`, `useStorageState`, `useTTS`, `useTokenCounters`, `useTouchGestures`), storage classes, utils, constants, defaults, and polyfills (104 files, +2894 / −1314 lines)
- **Server**: Convert server entry point, lib, and route modules from `.js` to `.ts`; switch to ESM (`"type": "module"`); add `tsconfig.json`; provide type declarations for `zstd` and `env`; update start scripts and Dockerfile for `tsx` runtime
- **Build**: Enable `strict` mode in the root `tsconfig.json` and install `typescript`, `tsx`, `@types/*` dev dependencies

### Type Safety Improvements
- **New type definition files**:
  - `src/types/api.d.ts` — API function signatures and response types
  - `src/types/components.d.ts` — Component prop interfaces
  - `src/types/contexts.d.ts` — Context value types for GenerationContext and SettingsContext
  - `src/types/defaults.d.ts` — Preset/template/theme shape types
  - `src/types/global.d.ts` — Global augmentations (window, etc.)
  - `src/types/storage.d.ts` — Storage class interfaces
- **Removed `as any` casts** across multiple subsystems: dispatch call sites, DOM access, file loading, object builder, data access, config, group filtering, generator chains (CompletionChunk/AIHordeChunk), BiasItem/BiasTempState state, and TokenizeResult
- **Added proper type annotations** to component props, hook parameters/variables, API function parameters, context providers, storage subclasses, and utility functions

### Strict Mode Fixes
- Resolved strict mode errors in: App, utils, storage classes, context providers, hooks, API modules, components, AppLayout, modals, worldinfo, and remaining files

## Impact
- No runtime behavior changes — the application compiles to identical JavaScript
- Enables future refactoring with full type-checking guardrails
- Lays groundwork for IDE autocompletion and better developer experience

## Help Wanted
This is a large migration touching every file in the project. I've verified that `npx tsc --noEmit` passes cleanly and the dev server starts, but thorough end-to-end testing is needed. If you can help test any of the following areas, please report issues or leave a comment:

- **Generation flows**: KoboldCPP, LlamaCPP, OpenAI, AI Horde, DeepSeek — try sending prompts and confirm responses appear correctly
- **Storage & sessions**: Create/switch/delete sessions, verify persistence across page reloads
- **Modals**: Open and interact with each modal (Settings, Connection Manager, Logit Bias, World Info, Instruct Templates, Theme Manager, etc.)
- **Server backend**: Run the Node.js server (`npm run server`) and test data routes, proxy, tokenizer, and backup
- **Token counters**: Confirm per-model token estimates look reasonable
- **Edge cases**: Empty states, rapid clicking, concurrent generations

Please open an issue or comment here with any bugs found. Thanks for helping shake out the remaining issues!

## Summary by CodeRabbit

* **New Features**
  * Hardened proxying: blocks private/internal targets and enforces safe final URLs.
  * Added Zstd-backed database compression with automatic migrations and scheduled maintenance.
  * Added server-backed tokenizer and token-counting/tokenization endpoints.
  * Added a server-backed storage adapter option alongside IndexedDB.

* **Improvements**
  * Improved Markdown preview safety by sanitizing rendered HTML.
  * Improved reliability with safer abort/error handling and more defensive migration behavior.
  * Updated server startup/entry to run the TypeScript server directly (and changed default bind host to `127.0.0.1`).
  * Updated README with clearer **Security** deployment guidance.

### CODE REVIEW

File: server/routes/proxy.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    return PRIVATE_IP_RANGES.some(r => num >= r.start && num <= r.end);
+}
+
+function isPrivateIPv6(ip: string): boolean {
+    const lower = ip.toLowerCase();
+    if (lower === '::1' || lower === '0:0:0:0:0:0:0:1') return true;
+    if (lower.startsWith('fc') || lower.startsWith('fd')) return true;
+    if (lower.startsWith('fe8') || lower.startsWith('fe9') || lower.startsWith('fea') || lower.startsWith('feb')) return true;
+    return false;
+}
```
[the remark this answers is not in the document]
Comment (LordFoogThe4rd) [reply server/routes/proxy.ts:34]: Out of scope, won’t fix.

File: src/utils/regex.ts [hunk 2]
```diff
@@ -41,40 +44,41 @@ export function regexLastIndexOf(string, regex, startpos) {
...
 		else {
-			let result = fn(n);
-			cache[n] = result;
+			let result = fn(...args);
+			cache[key] = result;
 			return result;
 		}
 	}
+	return memoized as T;
 }
```
[the remark this answers is not in the document]
Comment (LordFoogThe4rd) [reply src/utils/regex.ts:61]: For future me.

**Comment by LordFoogThe4rd** [comment]
Not fixing network security problems associated with public hosting anymore. It is your responsibility to secure your instance in that case.

File: server/routes/proxy.ts [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        if (isPrivateHostname(parsed.hostname)) return false;
+
+        const lookup = dns.lookup(parsed.hostname, { all: true });
+        const timeout = new Promise<never>((_, reject) =>
+            setTimeout(() => reject(new Error('DNS timeout')), 5000)
+        );
+        const addresses = await Promise.race([lookup, timeout]);
+        for (const addr of addresses) {
+            if (addr.family === 4 && isPrivateIP(addr.address)) return false;
+            if (addr.family === 6 && isPrivateIPv6(addr.address)) return false;
```
[the remark this answers is not in the document]
Comment (LordFoogThe4rd) [reply server/routes/proxy.ts:56]: We’ve already gone through this. Wontfix, PR welcome.

File: server/server.ts [hunk 4]
```diff
@@ -1,23 +1,31 @@
...
+import proxyRoutes from './routes/proxy.js';
+import zstdRoutes from './routes/zstd.js';
+import tokenizerRoutes from './routes/tokenizer.js';
+
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
+
+const args = minimist(process.argv.slice(2)) as Record<string, any>;
 const port = args.port || process.env.MIYAPAD_PORT || 3000;
-const host = args.host || process.env.MIYAPAD_HOST || '0.0.0.0';
+const host = args.host || process.env.MIYAPAD_HOST || '127.0.0.1';
```
[the remark this answers is not in the document]
Comment (LordFoogThe4rd) [reply server/server.ts:22]: Documentation will be updated later.
